### PR TITLE
add scoring methods in Luong-style attention

### LINKED ISSUE #15866: Scoring methods in Luong-style attention
TensorFlow version (you are using): 2.7.0
Are you willing to contribute it (Yes/No) : Yes

**Describe the feature and the current behavior/state**.

Describe the feature clearly here. Be sure to convey here why the requested feature is needed. Any brief description about the use-case would help.
Luong-style attention attention use three types of scoring methods, namely dot, general and concat. This can be found in the 3rd page of the [original paper](https://arxiv.org/pdf/1508.04025.pdf) and explained [here](https://stackoverflow.com/a/44454253). 
tiQkz

Right now `concat` method for scoring is not implemented and only [dot and general](https://github.com/keras-team/keras/blob/master/keras/layers/dense_attention.py#L339) has been implemented in the layer. An use case can be found [here](https://github.com/spro/practical-pytorch/blob/master/seq2seq-translation/seq2seq-translation.ipynb). 

**Will this change the current api? How?**
Yes. An extra parameter in the Attention layer to specify the scoring method. 

**Who will benefit from this feature?**
Anyone using Luong-style attention layer can opt for concat scoring method.

### PR DESCRIPTION
**Pull request by old-school-kid**
Luong-style attention attention use three types of scoring methods, namely dot, general and concat. This can be found in the 3rd page of the [original paper](https://arxiv.org/pdf/1508.04025.pdf) and explained [here](https://stackoverflow.com/a/44454253). 
Implementation of the Layer can be found in this [notebook](https://colab.research.google.com/drive/1lShkQNWZJJdaGmm205CppFjbLnOx0OD_?usp=sharing)

Auto closes #15866

### CODE REVIEW

File: keras/layers/dense_attention.py [hunk 1]
```diff
@@ -242,6 +242,8 @@ class Attention(BaseDenseAttention):
       Defaults to `False`.
     dropout: Float between 0 and 1. Fraction of the units to drop for the
       attention scores. Defaults to 0.0.
+    score: One of {'dot', 'concat'}. 'dot' refers to dot multiplication
```
Comment (fchollet) [inline keras/layers/dense_attention.py:245]: Please rename argument to "score_mode" for consistency with other API conventions in Keras

File: keras/layers/dense_attention.py [hunk 2]
```diff
@@ -319,9 +321,14 @@ class Attention(BaseDenseAttention):
   ```
   """
 
-  def __init__(self, use_scale=False, **kwargs):
+  def __init__(self, use_scale=False, score='dot', **kwargs):
     super(Attention, self).__init__(**kwargs)
     self.use_scale = use_scale
+    self.score_type= score
+    if self.score_type not in ['dot', 'concat']:
+      logging.warning(f'Score type {self.score_type} is unknown, '
```
Comment (fchollet) [inline keras/layers/dense_attention.py:329]: This should be a ValueError

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

File: keras/layers/dense_attention.py [hunk 3]
```diff
@@ -242,6 +242,8 @@ class Attention(BaseDenseAttention):
       Defaults to `False`.
     dropout: Float between 0 and 1. Fraction of the units to drop for the
       attention scores. Defaults to 0.0.
+    score: One of {'dot', 'concat'}. 'dot' refers to dot multiplication
+      of query and key. 'concat' refers to hyperbolic tangent of query and key.
```
[the remark this answers is not in the document]
Comment (old-school-kid) [reply keras/layers/dense_attention.py:246]: > It's tanh of the sum

Its not exactly tanh of the sum. [Key and query are concatenated](https://github.com/old-school-kid/keras/blob/master/keras/layers/dense_attention.py#L358) much like score in AdditiveAttention layer (Bahdanau-style attention). 

> Why is this called concat in this case?

In the [paper](https://arxiv.org/pdf/1508.04025.pdf) they call this score concat. 
tiQkz

**Comment by old-school-kid** [comment]
Did the requested changes.

File: keras/layers/dense_attention.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (fchollet) [reply keras/layers/dense_attention.py:246]: Sounds good then, but please improve the argument description. Currently the description seems rather confusing.

[the remark this answers is not in the document]
Comment (old-school-kid) [reply keras/layers/dense_attention.py:246]: Is this okay? 
"score_mode: One of {'dot', 'concat'}. 'dot' refers to dot product of query and key. 'concat' refers to hyperbolic tangent of query and key concatenated."

Do we need to describe the parameter score_type too?

File: keras/layers/dense_attention.py [hunk 4]
```diff
@@ -319,12 +321,19 @@ class Attention(BaseDenseAttention):
...
 
-  def __init__(self, use_scale=False, **kwargs):
+  def __init__(self, use_scale=False, score_mode='dot', **kwargs):
     super(Attention, self).__init__(**kwargs)
     self.use_scale = use_scale
+    self.score_mode= score_mode
+    if self.score_mode not in ['dot', 'concat']:
+      raise ValueError(
+                "Unknown score_mode. Acceptable values "
+                "are: ['dot', 'concat']"
```
Comment (fchollet) [inline keras/layers/dense_attention.py:331]: Add:

```
f"Received: score_mode={score_mode}"
```

Also make sure you use `'` and `"` consistently

**Review by fchollet — COMMENTED** [review]
Thanks for the update!

**Comment by old-school-kid** [comment]
@fchollet can you please take a look?
Thanks in advance

**Review by fchollet — APPROVED** [review]
Thanks for the contribution!

File: keras/layers/dense_attention_test.py [hunk 5]
```diff
@@ -203,6 +203,30 @@ def test_calculate_scores_multi_dim(self):
...
+  def test_calculate_scores_multi_dim_concat(self):
+    # Query tensor of shape [1, 2, 4]
+    q = np.array([[[1., 1.1, 1.2, 1.3], [2., 2.1, 2.2, 2.3]]], dtype=np.float32)
+    # Key tensor of shape [1, 3, 4]
+    k = np.array(
+        [[[1.5, 1.6, 1.7, 1.8], [2.5, 2.6, 2.7, 2.8], [3.5, 3.6, 3.7, 3.8]]],
+        dtype=np.float32)
+    attention_layer = dense_attention.Attention(score_mode='concat')
+    attention_layer.build(input_shape=([1, 2, 4], [1, 3, 4]))
+    actual = attention_layer._calculate_scores(query=q, key=k)
```
Comment (fchollet) [inline keras/layers/dense_attention_test.py:216]: Test is failing: https://source.cloud.google.com/results/invocations/acd8077d-49cf-4ad9-a164-2c37718c3a47/targets/keras%2Fgithub%2Fubuntu%2Fcpu%2Fpresubmit/log

You may need to do `actual = keras.backend.get_value(actual)`

File: keras/layers/dense_attention_test.py [hunk 6]
```diff
@@ -203,6 +203,30 @@ def test_calculate_scores_multi_dim(self):
...
+    attention_layer.attention_v = 1
```
Comment (fchollet) [inline keras/layers/dense_attention_test.py:217]: I notice that it isn't clear what `attention_v` means. Can you rename it to a fully spelled out variable name?

Comment (old-school-kid) [reply keras/layers/dense_attention_test.py:217]: Ah, so in `concat` score is the product of a learnable parameter (here `attention_v`) and the hyperbolic tangent of the concatenation of the query and key vectors. 

tiQkz
v right here is what I am calling `attention_v` and as usual Wa is the scaling parameter and ht and hs are query and key respcetively. 
A pytorch implementation of the same can be found [here](https://github.com/spro/practical-pytorch/blob/master/seq2seq-translation/seq2seq-translation.ipynb) in code block 11.

File: keras/layers/dense_attention_test.py (hunk 5 above)
Comment (old-school-kid) [reply keras/layers/dense_attention_test.py:216]: Is it because I am not initializing the parameter `attention_v` and only create it if `score_mode= 'concat' ` ? Because `score_mode='dot' ` is passing and the scale parameter is also initialized to None at the start itself.

Comment (fchollet) [reply keras/layers/dense_attention_test.py:216]: So what do you get when you replace it with `actual = keras.backend.get_value(actual)`?

File: keras/layers/dense_attention_test.py (hunk 6 above)
Comment (fchollet) [reply keras/layers/dense_attention_test.py:217]: Rename it to something that isn't an abbreviation, like `attention_weights`. See the Keras API style guide comments on naming: https://github.com/keras-team/governance/blob/master/keras_api_design_guidelines.md#carefully-weigh-whether-a-new-feature-should-be-included

Comment (old-school-kid) [reply keras/layers/dense_attention_test.py]: Renamed `attention_v` to `concat_score_weight`. Does that work?

File: keras/layers/dense_attention_test.py (hunk 5 above)
Comment (old-school-kid) [reply keras/layers/dense_attention_test.py:217]: To maintain the uniformity in the tests I havent used this yet. 
The error msg is : 

> Node: 'ReadVariableOp'
Could not find variable attention_v. This could mean that the variable has been deleted. In TF1, it can also mean the variable is uninitialized. Debug info: container=localhost, status error message=Container localhost does not exist. (Could not find resource: localhost/attention_v)
	 [[{{node ReadVariableOp}}]]
	 

So initialized `attention_v` (now `concat_score_weight`) to None just like `scale` is done.

Comment (fchollet) [reply keras/layers/dense_attention_test.py:217]: The error is unrelated to this. It will likely go away if you use `get_value`.

Comment (old-school-kid) [reply keras/layers/dense_attention_test.py]: Did the requested changes.

**Review by fchollet — APPROVED** [review]
Thanks for the update -- let's try merging again.

**Comment by old-school-kid** [comment]
@fchollet 
Extremely sorry for the inconvenience. The paramter name was set according to the inital naming system. Made the correction to the single test.

**Review by fchollet — APPROVED** [review]
